### ISSUE TITLE
Feature: Add PostgreSQL replica support

### ISSUE THREAD

**Issue by jeroenhabets** [zitadel/zitadel#9109]
### Preflight Checklist

- [x] I could not find a solution in the existing issues, docs, nor discussions
- [x] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Describe your problem

If Zitadel were to support a separate read-only database connection most of the load could be offloaded from the primary to one of the replicas in a PostgreSQL HA cluster. 

NOTE: Not all read-only queries can be moved to a replica, e.g. in a critical path where a read is closely followed by a write of some data. Even so, with other HA/clustered we've achieved major cost savings by sizing a 3 node cluster to only 1/2 sized nodes of a one node solutions.

### Describe your ideal solution

An MVP would allow to specify an `ReadOnlyPort` (or `ReplicaPort`), e.g. for my use case with https://github.com/vitabaks/autobase
Or more advanced under `Database:` add support for `postgres_readonly` (`postgres_replica`) besides `postgres`

https://zitadel.com/docs/self-hosting/manage/database#zitadel-with-postgres

### Version

v2.66.1

### Environment

Self-hosted

**Comment by muhlemmer** [comment]
We currently have an internal discussion about redesigning the database architecture. We will keep this suggestion in mind.

Note that if we provide access to RO replica's read queries might return outdated results and on an ops level you will have to deal with the eventual consistency. In our experience eventual consistency seems to upset tools like terraform.

**Comment by jeroenhabets** [comment]
@muhlemmer thanks for the update!

Regarding your Note. I tried to address that concern a priori (see quote below): For exactly this reason a certain subset of read queries, where this inconsistency may lead to issues, _must_ be directed to the primary. When we did that in a previous project, despite only the remainder outside of this subset going to the secondaries, we still measured substantial gains.
> NOTE: Not all read-only queries can be moved to a replica, e.g. in a critical path where a read is closely followed by a write of some data. Even so, with other HA/clustered we've achieved major cost savings by sizing a 3 node cluster to only 1/2 sized nodes of a one node solutions.

**Comment by fforootd** [comment]
Slightly related comment from me here 😁 

Would it not be a better approach to move more hot resources to a shared cache and try to take some pressure of the primary db... especially reads?

I was just today talking with @livio-a about this and we think adding more resources to the cache for transactional load would be great and then enable read replicas for async tasks that can work with some delay ... like notifications and scheduled task.

**Comment by jeroenhabets** [comment]
Thank you @fforootd! I was under the impression that was already in place with e.g. the Redis cache connector (internal/cache/connector/redis) so my feature request was more as a further optimization and to efficiently move towards a clustered PostgreSQL database..
But even without metrics, I still suspect you are correct that if more data could be cached that'd probably give the biggest boost to performance.

With "enable read replicas for async tasks that can work with some delay ... like notifications and scheduled task" then targeting the PostgeSQL secondaries, you mean?

With this feature request Zitadel would move towards 
| Zitadel Cluster |
| -------------- |
| <---> Redis (or should I say Valkey) Cluster |
| <---> PostgreSQL Cluster |

I like PostgreSQL primary-secondary clusters for PostgreSQL Clusters as they are much simpler than multi-primary. And if the secondaries see read-only load you don't need to provision 3x primary as all 3 nodes would see less load so you get clustering on the cheap.

**Comment by fforootd** [comment]
> I was under the impression that was already in place with e.g. the Redis cache connector

Yeah we utilise redis (and its derivatives) for some objects, but there is a lot of headroom to add more 😁 

> With "enable read replicas for async tasks that can work with some delay ... like notifications and scheduled task" then targeting the PostgeSQL secondaries, you mean?

Lets say there are some task in the background that need to read events, for example creating a metric, or reading events to sync into a slower storage, or sending informational mails (like when a password changed). These all could be read from a read-only replica to keep the primary happy.

> I like PostgreSQL primary-secondary clusters for PostgreSQL Clusters as they are much simpler than multi-primary.

I agree, to me that is where we can look at the read-only replica concept.

**Comment by elinashoko** [comment]
Thanks for filing this issue. As part of an ongoing backlog cleanup, we're closing bug reports and requests filed against ZITADEL v2.x, since that's several major versions behind current (v4.x, with v5 in progress).

If you're still seeing this on a current version, let us know in a comment with your updated version info (or reopen the issue) and we'll take another look.

This issue will close in 30 days if there's no further activity.

**Comment by elinashoko** [comment]
Thanks for filing this issue. As part of an ongoing backlog cleanup, we're closing issues that have had no activity in over 12 months.

If this is still relevant, let us know in a comment (or reopen the issue) and we'll take another look.

This issue will close in 30 days if there's no further activity.
